examples/pi-home-router/src/ipv4/mod.rs

```diff
@@ -0,0 +1,388 @@
+
+mod classify_dest_subnet;
+use cidr::Cidr;
+use cidr::Ipv4Cidr;
+use classify_dest_subnet::ByDestSubnet;
+use route_rs_runtime::processor::Identity;
+use std::net::Ipv4Addr;
+
+pub(crate) struct HandleIPv4 {
+    in_stream: Option<PacketStream<InterfaceAnnotated<Ipv4Packet>>>,
+    wan_ip: Option<Ipv4Addr>,
+    lan_ip: Option<Ipv4Addr>,
+    lan_subnet: Option<Ipv4Cidr>,
+}
+
+#[allow(dead_code)]
+impl HandleIPv4 {
+    pub(crate) fn new() -> Self {
+        HandleIPv4 {
+            in_stream: None,
+            wan_ip: None,
+            lan_ip: None,
+            lan_subnet: None,
+        }
+    }
+
+    pub(crate) fn wan_ip(self, wan_ip: Ipv4Addr) -> Self {
+        match self.wan_ip {
+            None => HandleIPv4 {
+                in_stream: self.in_stream,
+                wan_ip: Some(wan_ip),
+                lan_ip: self.lan_ip,
+                lan_subnet: self.lan_subnet,
+            },
+            Some(_) => panic!("HandleIPv4 takes only one wan_ip"),
+        }
+    }
+
+    pub(crate) fn lan_ip(self, lan_ip: Ipv4Addr) -> Self {
+        match self.lan_ip {
+            None => HandleIPv4 {
+                in_stream: self.in_stream,
+                wan_ip: self.wan_ip,
+                lan_ip: Some(lan_ip),
+                lan_subnet: self.lan_subnet,
+            },
+            Some(_) => panic!("HandleIPv4 takes only one lan_ip"),
+        }
+    }
+
+    pub(crate) fn lan_subnet(self, lan_subnet: Ipv4Cidr) -> Self {
+        match self.lan_subnet {
+            None => HandleIPv4 {
+                in_stream: self.in_stream,
+                wan_ip: self.wan_ip,
+                lan_ip: self.lan_ip,
+                lan_subnet: Some(lan_subnet),
+            },
+            Some(_) => panic!("HandleIPv4 takes only one lan_subnet"),
+        }
+    }
+}
+
+impl LinkBuilder<InterfaceAnnotated<Ipv4Packet>, InterfaceAnnotated<Ipv4Packet>> for HandleIPv4 {
+    fn ingressors(self, mut in_streams: Vec<PacketStream<InterfaceAnnotated<Ipv4Packet>>>) -> Self {
+        if self.in_stream.is_some() {
+            panic!("HandleIPv4 takes only one ingressor");
+        }
+        if in_streams.len() == 1 {
+            HandleIPv4 {
+                in_stream: Some(in_streams.remove(0)),
+                wan_ip: self.wan_ip,
+                lan_ip: self.lan_ip,
+                lan_subnet: self.lan_subnet,
+            }
+        } else {
+            panic!("HandleIPv4 takes exactly one ingressor");
+        }
+    }
+
+    fn ingressor(self, in_stream: PacketStream<InterfaceAnnotated<Ipv4Packet>>) -> Self {
+        match self.in_stream {
+            None => HandleIPv4 {
+                in_stream: Some(in_stream),
+                wan_ip: self.wan_ip,
+                lan_ip: self.lan_ip,
+                lan_subnet: self.lan_subnet,
+            },
+            Some(_) => panic!("HandleIPv4 takes only one ingressor"),
+        }
+    }
+
+    fn build_link(self) -> Link<InterfaceAnnotated<Ipv4Packet>> {
```

Comment:
I don't think this is actually that complex of a logic, there are maybe 20 links in here. The problem I think is that setting up each element is this huge block even though conceptually there's not that much going on.

Personally I don't like trying to split things into so many pieces just for the goal of splitting them up, because it can become much harder to follow the flow of what's happening.